j5/backends/hardware/sr/v4/serial/power_board.py

```diff
@@ -0,0 +1,298 @@
+from time import sleep
+from typing import Dict, Set, cast
+
+from serial.tools.list_ports_common import ListPortInfo
+
+from j5.backends import Backend, CommunicationError
+from j5.backends.hardware import (
+    DeviceMissingSerialNumberError,
+    NotSupportedByHardwareError,
+)
+from j5.backends.hardware.sr.v4.serial.protocol import (
+    SRV4SerialProtocolBackend,
+)
+from j5.boards import Board
+from j5.boards.sr.v4.power_board import PowerBoard
+from j5.components import (
+    BatterySensorInterface,
+    ButtonInterface,
+    LEDInterface,
+    PiezoInterface,
+    PowerOutputInterface,
+)
+
+LED_NAMES = ("RUN", "ERR")
+MAX_BUZZ_DURATION_MS = (2 ** 31) - 1  # int32 max
+MINIMUM_MAJOR_VERSION = 4
+
+
+def is_power_board(port: ListPortInfo) -> bool:
+    """
+    Check if a ListPortInfo represents a PBV4B.
+
+    We know that the Power Board must be running compatible firmware
+    as it must expose a serial interface to be listed.
+
+    :param port: ListPortInfo object.
+    :returns: True if object represents a power board.
+    """
+    return port.manufacturer == "Student Robotics" and port.product == "Power Board v4" \
+        and port.vid == 0x1bda and port.pid == 0x0010
+
+
+class SRV4SerialProtocolPowerBoardHardwareBackend(
+    SRV4SerialProtocolBackend,
+    PowerOutputInterface,
+    PiezoInterface,
+    ButtonInterface,
+    BatterySensorInterface,
+    LEDInterface,
+):
+    """The hardware implementation of the SR V4 power board using the serial protocol."""
+
+    board = PowerBoard
+
+    @classmethod
+    def discover(cls) -> Set[Board]:
+        """
+        Discover boards that this backend can control.
+
+        :returns: set of boards that this backend can control.
+        :raises DeviceMissingSerialNumberError: a board without a serial number was found.
+        """
+        # Find all serial ports.
+        ports = cls.get_comports()
+
+        # Get a list of boards from the ports.
+        boards: Set[Board] = set()
+        for port in filter(is_power_board, ports):
+            if port.serial_number is None:
+                raise DeviceMissingSerialNumberError(
+                    "Found power board-like device without serial number. "
+                    f"The power board is likely to be damaged: {port.usb_info()}",
+                )
+            else:
+                boards.add(
+                    PowerBoard(
+                        port.serial_number,
+                        cast(
+                            Backend,
+                            cls(port.device),
+                        ),
+                    ),
+                )
+
+        return boards
+
+    def __init__(
+            self,
+            serial_port: str,
+    ) -> None:
+        super().__init__(
+            serial_port=serial_port,
+            baud=115200,
+        )
+        self._lock = threading.Lock()
+        self._led_states: Dict[int, bool] = {
+            i: False
+            for i in range(2)
+        }
```

Comment:
Is this actually clearer than just explicitly creating the dict fo 0 & 1?

Reply:
e4c482968c25037946c69e16290c6a30ce2dd488

Reply:
This commit doesn't seem to be in this PR